packages/core/contracts/financial-templates/contract-for-difference/ContractForDifference.sol

```diff
@@ -0,0 +1,183 @@
+
+        // Get the current settlement price and store it. If it is not resolved will revert.
+        if (contractState != ContractState.ExpiredPriceReceived) {
+            expiryPrice = _getOraclePriceExpiration(expirationTimestamp);
+            expirationTokensForCollateral = financialProductLibrary.expirationTokensForCollateral(expiryPrice);
+            contractState = ContractState.ExpiredPriceReceived;
+        }
+
+        require(longToken.burnFrom(msg.sender, longTokensToRedeem));
+        require(shortToken.burnFrom(msg.sender, shortTokensToRedeem));
+
+        // expirationTokensForCollateral is a number between 0 and 1e18. 0 means all collateral goes to short tokens and
+        // 1 means all collateral goes to the long token. Total collateral returned is the sum of payouts.
+        uint256 collateralPerToken = collateralPerPair.div(2);
+
+        uint256 longCollateralRedeemed =
+            FixedPoint
+                .fromUnscaledUint(longTokensToRedeem)
+                .mul(collateralPerToken)
+                .mul(expirationTokensForCollateral)
+                .rawValue;
+        uint256 shortCollateralRedeemed =
+            FixedPoint
+                .fromUnscaledUint(shortTokensToRedeem)
+                .mul(collateralPerToken)
+                .mul(FixedPoint.fromUnscaledUint(1).sub(expirationTokensForCollateral))
+                .rawValue;
+
+        collateralReturned = longCollateralRedeemed.add(shortCollateralRedeemed);
+        collateralToken.safeTransfer(msg.sender, collateralReturned);
+    }
+
+    function _getOraclePriceExpiration(uint256 requestedTime) internal returns (int256) {
+        // Create an instance of the oracle and get the price. If the price is not resolved revert.
+        OptimisticOracleInterface optimisticOracle = _getOptimisticOracle();
+        require(optimisticOracle.hasPrice(address(this), priceIdentifier, requestedTime, _getAncillaryData()));
+        return optimisticOracle.settleAndGetPrice(priceIdentifier, requestedTime, _getAncillaryData());
+    }
+
+    function _requestOraclePriceExpiration() internal {
+        OptimisticOracleInterface optimisticOracle = _getOptimisticOracle();
+
+        // For now, we add no fees the the OO and set the reward to 0.
+        optimisticOracle.requestPrice(priceIdentifier, expirationTimestamp, _getAncillaryData(), collateralToken, 0);
+    }
+
+    function _getIdentifierWhitelist() internal view returns (IdentifierWhitelistInterface) {
+        return IdentifierWhitelistInterface(finder.getImplementationAddress(OracleInterfaces.IdentifierWhitelist));
+    }
+
+    function _getOptimisticOracle() internal view returns (OptimisticOracleInterface) {
+        return OptimisticOracleInterface(finder.getImplementationAddress(OracleInterfaces.OptimisticOracle));
+    }
+
+    function _getAncillaryData() internal view returns (bytes memory) {
+        return abi.encodePacked(address(longToken), address(shortToken));
+    }
```

Comment:
We don't need to do this in this PR, but do you think we should allow the deployer to customize the ancillary data?

Reply:
sure. For now I left it like this just to follow the same pattern as the EMP. we totally should change this in the future.